### PR TITLE
Media Library: try grouping by date (React Virtualized)

### PR DESCRIPTION
**Pull request by iseulde**
This PR is similar to #10238, but it **builds on a media library that uses React Virtualized**. See #10285 for more on that.

Branch diff: https://github.com/Automattic/wp-calypso/compare/update/media-library-react-virtualized...try/media-library-date-groups-virtualized

It only groups the media by date, displays the date at the top (wrapped in an `h3` element) in a pretty format ("Today", "Yesterday", day of the week if within 7 days, day and month if within a year, and a full date for the rest). Each group starts on a new line. ~~It has no further styling.~~

screen shot 2017-01-05 at 12 36 17

The original mock up from @adambbecker has smaller groups aligned:

zoomed-default_grid

I left a comment regarding that (p5j4vm-1rz-p2):

> I’ve been thinking a bit more about the aligning of images, as it does seem to be in scope of the task. It would indeed be good to align small groups of items together. A disadvantage, though, is that it becomes difficult to scan through, which is why we’re adding these dates in the first place. I think it is best to keep the group label (i.e. the dates) left aligned (and right for RTL languages of course).
>
> An alternative would be to group smaller groups together and have the label be a range of dates. The question is when to start and when to stop. If you have mostly small groups, or the range is big, you might end up with ranges like 17 May – 8 November. To prevent this we could draw the line at the months. This might still look weird if you don’t have much media, which makes me wonder if we should remove the date grouping if there are too few images.
>
> Like this is becomes quite complex, and I’m not sure at all if it’s the right way to go. It’s very unlike a media library of a camera; I think most people just use a few media per post, but I lack the data to know.
>
> Another alternative could be to group the media in “smart” albums and only display the first, which would then link to a list of all the images uploaded that day. Single images would not link to a list and are “loose”, but as soon as there’s more than one uploaded on the same day, it has a list. Like this there will be no gaps at all in he layout, it’s easy to scan, and each tile has a label with the day. This can even be done for other things like the ordering by uploaded post.
>
> I would love to get some feedback on these issues. For now I’ll just leave them grouped by day, starting on a new line.

And:

> I think the “smart” album approach might work quite well if you have tons of media too. When I upload 100+ pictures of an event that I want to link, I’d no longer have to scroll through all of them before I reach older pictures, which is one of the most frustrating bits about the media library. Not only do you have to scroll through all of them, but you have to wait as they load, 20 at the time.

### CODE REVIEW

**Comment by folletto** [comment]
I agree with your analysis above. 

> I think most people just use a few media per post, but I lack the data to know.

Same here. I feel the current media library is missing a "This Post / Page" filter.

> For now I’ll just leave them grouped by day, starting on a new line.

Looks like a good approach to start with.

I'm not sure on the level of the design review since it's still in progress, but just in case:

(1) I'd tweak the font. It's good to have a higher contrast for accessibility, but I feel it's a bit too big right now (the example below is `11px` and `darken( $gray, 20% )`). 

screen shot 2017-01-05 at 14 00 59


(2) I'm getting a horizontal scrollbar:

screen shot 2017-01-05 at 13 55 40

**Comment by iseulde** [comment]
> Do you have the "scrollbars always on" setting active?

I had no idea this existed! Thanks :)

**Comment by iseulde** [comment]
Digging into React Virtualized to figure out what's causing it.

**Comment by iseulde** [comment]
Updated the parent PR rebased this one.

Any changes specific to this PR can be seen here: https://github.com/Automattic/wp-calypso/compare/update/media-library-react-virtualized...try/media-library-date-groups-virtualized

**Comment by iseulde** [comment]
@folletto 11px is pretty small for my eyes. I see it is use elsewhere though for small labels. Could it maybe be a little bit bigger?

**Comment by folletto** [comment]
> 11px is pretty small for my eyes. I see it is use elsewhere though for small labels. Could it maybe be a little bit bigger?

I've mentioned 11px since it's the same size of other kinds of similar headers (even if we use all caps there - see Pages or Notification subheaders) so I'd prefer to stick with that, unless we start changing all of them.

However, we also use uppercase... so maybe we can do 11px uppercase or 12px normal case? What do you think?

**Comment by iseulde** [comment]
Yes, upcast it looks a bit better. :) Updated.

**Comment by iseulde** [comment]
screen shot 2017-01-05 at 16 46 24

I added a little loading placeholder for the date too, so it all flows nicely when the first batch arrives.

**Comment by iseulde** [comment]
screen shot 2017-01-05 at 16 49 14

And this is the current look.

**Comment by folletto** [comment]
Multiple 👍  👍  👍  from me.

**Comment by iseulde** [comment]
The `D MMMM` format used in moment will need to be localised bit not sure where this is done or mapped.

File: client/my-sites/media-library/list.jsx [hunk 1]
```diff
@@ -182,61 +182,183 @@ export const MediaLibraryList = React.createClass( {
...
 
+		return moment.format( 'LL' );
+	},
+
+	getGridItems: function() {
+		const items = [];
+
+		this.props.media.forEach( ( item, index, media ) => {
+			if ( ! index || ! this.props.moment( media[ index - 1 ].date ).isSame( item.date, 'day' ) ) {
+				const trailing = items.length % this._columnCount;
```
Comment (ehg) [inline client/my-sites/media-library/list.jsx]: It's not completely clear as to what `trailing` means here.

File: client/my-sites/media-library/list.jsx [hunk 2]
```diff
@@ -182,61 +182,183 @@ export const MediaLibraryList = React.createClass( {
...
+				} );
+
+				items.push( ...Array( this._columnCount - 1 ) );
+			}
+
+			items.push( item );
+		} );
+
+		if ( ! this.props.media.length ) {
+			items.push( {
```
Comment (ehg) [inline client/my-sites/media-library/list.jsx]: Why do we need to push an empty heading if we have no media?

**Comment by ehg** [comment]
Should we add some unit tests for heading functionality here?

File: client/my-sites/media-library/list.jsx (hunk 1 above)
Comment (iseulde) [inline client/my-sites/media-library/list.jsx]: I'll leave a comment or adjust the name. Trailing empty slots to be filled up.

File: client/my-sites/media-library/list.jsx (hunk 2 above)
Comment (iseulde) [inline client/my-sites/media-library/list.jsx]: To display a placeholder while loading the first batch of media. I'll add a comment.

**Comment by iseulde** [comment]
Yes, the method that puts all the grid items together could benefit from some tests.

**Comment by iseulde** [comment]
`formatDate` too actually. Just some dates and expected strings.

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**